src/NLog/Layouts/Typed/Layout.cs

```diff
@@ -0,0 +1,337 @@
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE 
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE 
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF 
+// THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+using System;
+using System.ComponentModel;
+using System.Globalization;
+using System.Reflection;
+using NLog.Common;
+
+namespace NLog.Layouts
+{
+    /// <summary>
+    /// Layout with a simple value (e.g. int) or a layout which results in a simple value (e.g. ${counter})
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+  //todo  [CLSCompliant(false)] // IConvertible
+    public class Layout<T> : TypedLayoutBase<T>
+  //  where T: IConvertible //todo not CLS compliant
+    {
+        private static TypeCode? _typeCode; 
+      //todo  private static bool _isEnum;
+
+
+        /// <inheritdoc />
+        public Layout(T value) : base(value)
+        {
+        }
+
+        /// <inheritdoc />
+        public Layout(Layout layout) : base(layout)
+        {
+        }
+
+        static Layout()
+        {
+            _typeCode = (default(T) as IConvertible)?.GetTypeCode();
+           //todo  _isEnum = typeof(T).IsEnum;
+        }
+
+        #region Overrides of TypedLayout<T>
+
+        /// <inheritdoc />
+        protected override string TypedName => typeof(T).Name; //todo cache?
+
+        /// <inheritdoc />
+        protected override string ValueToString(T value, CultureInfo cultureInfo)
+        {
+            if (value is IConvertible convertible)
+            {
+                return convertible.ToString(cultureInfo);
+            }
+
+            return value?.ToString();
+        }
+
+        /// <inheritdoc />
+        protected override bool TryParse(string text, out T value)
+        {
+            return TryConvertTo(text, out value);
+        }
+
+        /// <inheritdoc />
+        protected override bool TryConvertTo(object raw, out T value)
+        {
+            value = default; // todo check
+            if (raw == null)
+            {
+                return true;
+            }
+
+            try
+            {
+                // todo could be better?
+                var objTypeCode = _typeCode;
+                switch (objTypeCode)
+                {
+                    //todo ugly :P
+                    case TypeCode.Boolean:
+                        {
+                            if (bool.TryParse(raw.ToString(), out var value1))
+                            {
+                                value = (T)(object)value1;
```

Comment:
dunno how to fix this box-unbox operation. If T is a bool, then this is really meh